app/mqtt_reader.py

```diff
@@ -0,0 +1,86 @@
+"""MQTT Reader module.
+
+This module provides a client class for subscribing to topics and receiving messages from MQTT brokers.
+
+"""
+
+import logging
+from typing import Callable
+
+import paho.mqtt.client as mqtt
+
+from app.message import CommandMessage
+from app.exceptions import InvalidMessageError
+from app.error_handler import ErrorHandler
+from app.mqtt_client import MqttClient
+
+
+def _decode_message(message):
+    return message.payload.decode()
+
+
+class MqttReader(MqttClient):
```

Comment:
Whats the need to use inheritance over composition here?
I don't think we need to think of the reader and writer of siblings.

I would instead recommend to think of them as 2 distinct actions, that happen to depend on the same dependency upon instantiation. But that's encapsulated

Reply:
I've refactored to have MQTTReader and MQTTWriter instantiate an MQTTClient class which they use to connect (I stored it in `self.connector` rather than `self.client` because the multiple levels of "client" were confusing). But I actually don't think it's worth it - the only useful thing left in the class is a single `connect` method and I think repeating that within reader and writer classes is an acceptable level of duplication.

Reply:
yes, that's fine. We can inject the same instance. In this case, paho remains an implementation detail of each object. There is not exposure of the MQTT interface to the clients of the reader or error handler.
We can swap or change things in the future and clients shouldn't care about such changes.

Reply:
OK that's done, MQTTClient is removed and both reader and writer have their own connect method.

And it may be asking for trouble, but I reworked ErrorHandler to use a separate class when it's using MQTT to transmit errors. The previous implementation required several optional args to `init` because the MQTT error handling may or may not be activated. I think this way is clearer, but happy to adjust if inheritance is not appropriate here either :)